Add AzureOIDC provider

thereallukl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.

Commit:
add pagination support for /user/teams

Commit:
limit => per_page

Commit:
change per_page value from 200 to 100

Ramblurr committed:
Add Nextcloud provider (#179)

thereallukl committed:
Add AzureOIDC provider

eerichmond commented on providers/azure_oidc.go:
```diff
@@ -0,0 +1,112 @@
+		OIDCProvider: NewOIDCProvider(p),
+	}
+	provider.ProviderName = "Azure OpenID Connect"
+	provider.ApprovalPrompt = "consent"
```
Can you make the ApprovalPrompt come from the input param instead of hardcoding as "consent"

eerichmond commented on providers/azure_oidc.go:
```diff
@@ -0,0 +1,112 @@
+	}
+	provider.ProviderName = "Azure OpenID Connect"
+	provider.ApprovalPrompt = "consent"
+	p.Scope = "openid email profile"
```
Scope is being hardcoded here as well as set inside of NewOIDCProvider.

eerichmond commented on providers/azure_oidc.go:
```diff
@@ -0,0 +1,112 @@
+	var claims AzureOIDCClaims
+	err := getClaimFromToken(session.IDToken, &claims)
+	if err != nil {
+		logger.Printf("error: failed to parse IDToken")
```
Can you include the original err in the logger.Printf?

cgroschupp committed:
Add static upstream

cgroschupp committed:
Add tests for static upstream

eerichmond commented on main.go:
```diff
@@ -25,6 +25,8 @@ func main() {
 	skipAuthRegex := StringArray{}
 	jwtIssuers := StringArray{}
 	googleGroups := StringArray{}
+	azurePermittedGroups := StringArray{}
```
Maybe shorten this to just azureGroups to match googleGroups above it

eerichmond commented:
I would love to see this merged into pusher/oauth2_proxy. Let me know if you want me to contribute to the message that goes into the pull request.

toshi-miura committed:
Merge branch 'ap-gh-paginate' of https://github.com/apratina/oauth2_proxy into ap-gh-pagination-with-lastpage

cgroschupp committed:
Move responceCode out of HandleFunc.

cgroschupp committed:
replace getRootEndpoint by getEndpointWithCookie

toshi-miura committed:
Added handling of link header in githubAPI paging process

======================================================
changelog note

[#274](https://github.com/pusher/oauth2_proxy/pull/274)  Add github api pagination support (@toshi-miura ,@apratina)

======================================================

I didn't edit CHANGELOG.md.
Since # 102 was taken over and the change difference of CHANGELOG.md was large

eerichmond commented:
Have you given up on this fork? Should I fork it and maintain it myself?

cgroschupp committed:
Update changelog

toshi-miura committed:
ReadAll() & Close() 　close together.

toshi-miura committed:
add change log.

JoelSpeed committed:
Merge branch 'master' into ap-gh-pagination-with-lastpage

kamaln7 committed:
allow redirects to whitelisted hosts with ports

toshi-miura committed:
Fix typo.

kamaln7 committed:
implicit/explicit redirect port matching

toshi-miura committed:
Update CHANGELOG.md

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

JoelSpeed committed:
Merge branch 'master' into ap-gh-pagination-with-lastpage

kamaln7 committed:
update port whitelisting rules, refactor IsValidRedirect tests

kamaln7 committed:
Merge branch 'master' into kamal/whitelist-redirects-with-ports

kamaln7 committed:
Update documentation and changelog

toshi-miura committed:
Merge branch 'master' into ap-gh-pagination-with-lastpage

cgroschupp committed:
Merge branch 'master' into feat/static-upstream

thereallukl committed:
Add support for Redis with custom CA.

thereallukl committed:
User preferred_username (https://docs.microsoft.com/pl-pl/azure/active-directory/develop/id-tokens) as username when using azure oidc

thereallukl committed:
Add docs

thereallukl committed:
Code formatting. Add missing CHANGELOG entry.

Commit:
Update main.go

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

thereallukl committed:
log message in case of failure during loading system cert pool

thereallukl committed:
Fix settings naming

JoelSpeed committed:
Merge pull request #309 from lleszczu/customCaForRedis

Add support for Redis with custom CA.

JoelSpeed committed:
Merge branch 'master' into ap-gh-pagination-with-lastpage

JoelSpeed committed:
Merge pull request #274 from toshi-miura/ap-gh-pagination-with-lastpage

Added pagination support for /user/teams github api with link header .

kamaln7 committed:
remove unnecessary if conditional

kamaln7 committed:
Merge branch 'master' into kamal/whitelist-redirects-with-ports

JoelSpeed committed:
Fixup Changelog entries (#312)

cgroschupp committed:
Merge branch 'master' into feat/static-upstream

JoelSpeed committed:
Merge pull request #265 from cgroschupp/feat/static-upstream

Add upstream with static response

morarucostel committed:
Add redirect capability to sign_out (#314)

* addint redirect capability to sign_out

* updating changelog

Commit:
Allow to change provider's name (#296)

* Allow to change provider's name.

* Add changelog entry.

* Linting.

* provider-name -> provider-display-name.

* Add flag in main.go.

* Update CHANGELOG.md

fnkr committed:
dist.sh: Add build for FreeBSD (#317)

andymelichar commented:
Trying this out for myself... how are you handling the email_verified from Azure? It doesn't look like it's part of the default payload and so I'm getting back "Error redeeming code during OAuth2 callback: email <myemailaddress> not listed as verified"

Commit:
Update CHANGELOG for v4.1.0 (#323)

steakunderscore committed:
Add notes on how to create a release

Commit:
dist.sh: use sha256sum (#325)

* dist.sh: use sha256sum

* update CHANGELOG

steakunderscore committed:
Merge branch 'master' into release-notes

steakunderscore committed:
Update RELEASE.md

steakunderscore committed:
Update README.md

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

steakunderscore committed:
Update RELEASE.md

Commit:
Add configuration for cookie 'SameSite' value.

Values of 'lax' and 'strict' can improve and mitigate
some categories of cross-site traffic tampering.

Given that the nature of this proxy is often to proxy
private tools, this is useful to take advantage of.

See: https://www.owasp.org/index.php/SameSite

JoelSpeed committed:
Merge pull request #326 from thought-machine/release-notes

Add notes on how to create a release

JoelSpeed committed:
Merge branch 'master' into nextcloud-provider

JoelSpeed committed:
Merge pull request #179 from Ramblurr/nextcloud-provider

Add nextcloud provider

JoelSpeed committed:
Fix changelog entry release

JoelSpeed committed:
Merge pull request #332 from pusher/fix-changelog

Fix changelog entry release

Commit:
Add item to CHANGELOG for SameSite configuration.

Commit:
Fix typos in doc strings.

kamaln7 committed:
Merge branch 'master' into kamal/whitelist-redirects-with-ports

sushiMix committed:
Update keycloak provider configuration doc (#347)

* update keycloak provider configuration doc

* Add changelog entry

Commit:
Merge branch 'master' into kamal/whitelist-redirects-with-ports

thereallukl committed:
Sort imports. Close request body

kamaln7 committed:
DigitalOcean Auth Provider (#351)

* DigitalOcean provider

* documentation: digitalocean provider

* changelog: digitalocean provider

* codeowners: digitalocean provider

Commit:
Merge branch 'master' into kamal/whitelist-redirects-with-ports

JoelSpeed committed:
Merge pull request #280 from kamaln7/kamal/whitelist-redirects-with-ports

Allow redirects to whitelisted hosts with ports

JoelSpeed committed:
Merge branch 'master' into pgroudas/add-samesite-cookie-options

JoelSpeed committed:
Merge pull request #339 from clubhouse/pgroudas/add-samesite-cookie-options

Add SameSite cookie configuration value for session cookie

dependabot[bot] committed:
Bump rubyzip from 1.2.2 to 2.0.0 in /docs

Bumps [rubyzip](https://github.com/rubyzip/rubyzip) from 1.2.2 to 2.0.0.
- [Release notes](https://github.com/rubyzip/rubyzip/releases)
- [Changelog](https://github.com/rubyzip/rubyzip/blob/master/Changelog.md)
- [Commits](https://github.com/rubyzip/rubyzip/compare/v1.2.2...v2.0.0)

Signed-off-by: dependabot[bot] <support@github.com>

JoelSpeed committed:
Merge pull request #357 from pusher/dependabot/bundler/docs/rubyzip-2.0.0

Bump rubyzip from 1.2.2 to 2.0.0 in /docs

JoelSpeed committed:
Add JoelSpeed to CODEOWNERS

Commit:
Fix linting error about const

JoelSpeed committed:
Merge pull request #361 from nawa/fix-linting

Fix linting error about const

JoelSpeed committed:
Merge pull request #358 from pusher/JoelSpeed-codeowners

Add @JoelSpeed to CODEOWNERS

martin-css committed:
Add reverse proxy setting (#331)

* Add reverse proxy setting (#321)

Commit:
check for /\ redirects

tomelliff committed:
Disable cgo on release builds

Tries to link against libc which isn't present in musl based distributions (such as Alpine Linux).

Closes https://github.com/pusher/oauth2_proxy/issues/329

Commit:
Merge pull request #365 from tomelliff/remove-cgo-from-release-builds

Disable cgo on release builds, will add this to the changelog for next release

Commit:
Release v5.0.0

Changelog updated

Commit:
Merge pull request #367 from pusher/feature/changelog-release-5

Release 5.0.0 / changelog update

Commit:
Merge branch 'master' into advisory-fix-1

Commit:
note about open redirect vulneravility

starkers committed:
Merge pull request from GHSA-qqxw-m5fj-f7gv

check for /\ redirects

Commit:
Open redirect (security vulnerability) notes

Commit:
Merge pull request #368 from pusher/advisory-notes

Open redirect (security vulnerability) notes

Commit:
Fix login page fragment handling after soft reload on Firefox (#353)

* Fix login page fragment handling after soft reload on Firefox.

* Add comments.

* Move changelog entry to correct place.

Co-authored-by: Joel Speed <Joel.speed@hotmail.co.uk>
Co-authored-by: Felix Fontein <felix@fontein.de>

holyjak committed:
lates version is v5 not v4 (#379)

nawatts committed:
Add sign_out endpoint to endpoint documentation page (#383)

yan-dblinf committed:
Extension of Redis Session Store to Support Redis Cluster (#363)

* Extend the redis session store to support redis cluster

* rename function newRedisClient to newRedisCmdable

* update docs about redis cluster as session store

* update autocomplete script with redis cluster options

* add check about conflict between option redis-use-sentinel and redis-use-cluster

* update change log

* Update docs/configuration/sessions.md

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

* Update pkg/sessions/redis/redis_store.go

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

* add the dropped option back

Co-authored-by: Joel Speed <Joel.speed@hotmail.co.uk>

howzat committed:
Support optional id_tokens in refresh responses (#335)

* OIDC Token Refresh works without id_tokens

Addresses https://github.com/pusher/oauth2_proxy/issues/318

Refactoring the OIDC provider so that the refresh process works when there are
no id_tokens present in the response. Added unit tests to the oidc_test.go to prove
the redeem and refresh still work.

The expiry time of the session is now taken from the outh token expiry and not
the id_token (preventing stale access_tokens in sessions).

* Refactoring the to use a KeySetStub in the oidc_test.go. This allows the
elimination of the slightly contrived function passing elements used
previously. (This change is being applied to address the bug #318)

* Changes as per the PR comments and preparing for 5.x release

* Fixup changelog

Co-authored-by: Joel Speed <Joel.speed@hotmail.co.uk>

Commit:
Update main.go

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

dmnemec committed:
GitHub Provider - Fallback to secondary verified email (#372)

* Fallback to secondary verified email

* Add change to changelog

pasha-r committed:
Support for client secret file. (#355)

* added ClientSecretFile in ProviderData

* add documentation notes on client secret file

* added Changelog entry for Client Secret File PR

* fixing configuration.md

* addressing PR issue of ClientSecret property naming

* Update providers/provider_data.go

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

* corrected changelog entry

* fixed typo in GetClientSecret

Co-authored-by: Joel Speed <Joel.speed@hotmail.co.uk>

Commit:
Fix #384 - link to sessions.md (#391)

Co-authored-by: Joel Speed <Joel.speed@hotmail.co.uk>

iain-buclaw-sociomantic committed:
Implement ValidateSessionState for GitHubProvider (#385)

Refactors the setting of the Authorization header into getGitHubHeader.

Refs #382

Co-authored-by: Joel Speed <Joel.speed@hotmail.co.uk>

niraizik committed:
Add client-secret-file to main FlagSet (#399)

thereallukl committed:
Fix formatting. Panic on failure to parse query in azure oidc provider.

Commit:
Docs: Fix link to sessions.md (#403)

luketainton committed:
Add nsswitch.conf to Docker image (#400)

* Add nsswitch.conf to Docker image

Created nsswitch.conf to use locally defined translations before DNS. Copied to /etc/nsswitch.conf in the image.

* Add new line

* Updated Changelog

Co-authored-by: Dan Bond <danbond@protonmail.com>

thereallukl committed:
AzureOIDC: Use /oidc/userinfo endpoint to read user profile. Improve docs.

thereallukl committed:
Merge remote-tracking branch 'upstream/master' into azure-oidc

Commit:
Update docs/2_auth.md

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

Commit:
Update docs/2_auth.md

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

Commit:
Update main.go

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>

Commit:
Update docs/2_auth.md

Co-Authored-By: Joel Speed <Joel.speed@hotmail.co.uk>
